docs/architecture/adr-033-protobuf-inter-module-comm.md

```diff
@@ -0,0 +1,341 @@
+
+```go
+package bank
+
+type MsgClient interface {
+	Send(context.Context, *MsgSend, opts ...grpc.CallOption) (*MsgSendResponse, error)
+}
+
+type MsgServer interface {
+	Send(context.Context, *MsgSend) (*MsgSendResponse, error)
+}
+```
+
+[ADR 021](./adr-021-protobuf-query-encoding.md) and [ADR 31](./adr-031-msg-service.md) specifies how modules can implement the generated `QueryServer`
+and `MsgServer` interfaces as replacements for the legacy queriers and `Msg` handlers respectively.
+
+In this ADR we explain how modules can make queries and send `Msg`s to other modules using the generated `QueryClient`
+and `MsgClient` interfaces and propose this mechanism as a replacement for the existing `Keeper` paradigm.
+
+Using this `QueryClient`/`MsgClient` approach has the following key benefits over keepers:
+1. Protobuf types are checked for breaking changes using [buf](https://buf.build/docs/breaking-overview) and because of
+the way protobuf is designed this will give us strong backwards compatibility guarantees while allowing for forward
+evolution.
+2. The separation between the client and server interfaces will allow us to insert permission checking code in between
+the two which checks if one module is authorized to send the specified `Msg` to the other module providing a proper
+object capability system.
+
+This mechanism has the added benefits of:
+- reducing boilerplate through code generation, and
+- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC 
+
+### Inter-module Communication
+
+In order for code to use the `Client` interfaces generated by the protobuf compiler, a `grpc.ClientConn` implementation
+is needed. We introduce a new type, `ModuleKey`, to serve this role which we can conceptualize as the "private key"
+corresponding to a module account.
+
+Whereas external clients use their private key to sign transactions containing `Msg`s where they are listed as signers,
+modules use their `ModuleKey` to send `Msg`s where they are listed as the sole signer to other modules. For example, modules
+could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another
+account.
+
+`QueryClient`s could also be made with `ModuleKey`s, except that authentication isn't required.
+
+Here's an example of a hypothetical module `foo` interacting with `x/bank`:
+```go
+package foo
+
+func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
+  bankQueryClient := bank.NewQueryClient(fooMsgServer.moduleKey)
+  balance, err := bankQueryClient.Balance(&bank.QueryBalanceRequest{Address: fooMsgServer.moduleKey.Address(), Denom: "foo"})
+  
+  ...
+
+  bankMsgClient := bank.NewMsgClient(fooMsgServer.moduleKey)
+  res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: fooMsgServer.moduleKey.Address(), ...})
+
+  ...
+}
+```
+
+### `ModuleKey`s and `ModuleID`s
```

Comment:
Is this `ModuleKey` intended to be used only for permissioning module accounts and sub-module accounts. Or could I use it to express arbitrary permissions?

For example, I want a supply module that provisions minting/burning permissions for different modules. So my supply module allows minting permission to ibc for any denomination that is prefixed by `ibc/` and nothing else, and no other module is allowed to mint denominations with `ibc/`.

That way I can ensure that no other module can mint `ibc/` tokens.

The current capability system is able to do things like this. I would prefer that we have a single dynamic inter-module capability system in the SDK. I have no strong opinion on whether it should be what is implemented in `x/capability` or this. But ideally we have a single system that supports all of our usecases, rather than multiple systems that support different subsets

Reply:
That could be accomplished with this system and is one of the goals laid out in #7093.

Reply:
The specifics of doing minting by denom prefix of course needs its own permission setup (as it would with either solution). But it's definitely a use case I'm holding in mind.